qiskit/transpiler/passes/layout/disjoint_utils.py

```diff
@@ -0,0 +1,117 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2023.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+
+"""This module contains common utils for disjoint coupling maps."""
+
+from collections import defaultdict
+from typing import List, Callable, TypeVar, Dict
+
+import rustworkx as rx
+
+from qiskit.circuit import Qubit
+from qiskit.dagcircuit.dagcircuit import DAGCircuit
+from qiskit.dagcircuit.dagnode import DAGInNode, DAGOpNode, DAGOutNode
+from qiskit.transpiler.coupling import CouplingMap
+from qiskit.transpiler.exceptions import TranspilerError
+
+T = TypeVar("T")
+
+
+def run_pass_over_connected_components(
+    dag: DAGCircuit,
+    coupling_map: CouplingMap,
+    run_func: Callable[[DAGCircuit, CouplingMap], T],
+    remove_barrier: bool = True,
+) -> List[T]:
+    """Run a transpiler pass inner function over mapped components."""
+    cmap_components = coupling_map.connected_components()
+    # If graph is connected we only need to run the pass once
+    if len(cmap_components) == 1:
+        return [run_func(dag, cmap_components[0])]
+    dag_components = separate_dag(dag, remove_barrier)
+    mapped_components = map_components(dag_components, cmap_components)
+    out_component_pairs = []
+    for cmap_index, dags in mapped_components.items():
+        out_dag = dag_components[dags.pop()]
+        for dag_index in dags:
+            dag = dag_components[dag_index]
+            out_dag.add_qubits(dag.qubits)
+            out_dag.add_clbits(dag.clbits)
+            out_dag.add_qreg(dag.qregs)
+            out_dag.add_clbits(dag.cregs)
+            out_dag.compose(dag, qubits=dag.qubits, clbits=dag.clbits)
+        out_component_pairs.append((out_dag, cmap_components[cmap_index]))
+    res = [run_func(out_dag, cmap) for out_dag, cmap in out_component_pairs]
+    return res
+
+
+def map_components(
+    dag_components: List[DAGCircuit], cmap_components: List[CouplingMap]
+) -> Dict[int, int]:
+    """Map a list of circuit components to coupling map components."""
+    free_qubits = {index: len(cmap.graph) for index, cmap in enumerate(cmap_components)}
+    out_mapping = defaultdict(list)
+
+    for dag_index, dag in enumerate(sorted(dag_components, key=lambda x: x.num_qubits())):
+        for cmap_index in range(len(cmap_components)):
```

Comment:
We should be also able to sort both `dag_components` and `cmap_components` by size (from largest to smallest) and match the `i`th `dag_component` to the `i`th `cmap_component`.

Reply:
I'm not sure we want to do the zip/1:1 mapping like this exactly because the dag components might be small enough that we need to put > 1 dag component on a single cmap component. But, my intent here was to sort the dag components from largest to smallest (which I realize looking at this now I did backwards, I need to add `reverse=True` to the `sorted()` call) so that we're getting a similar effect.

Reply:
Thanks! I did not realize that we might want to put multiple dag components on a single cmap component. Though would would happen in the following case? Let's suppose that the two dag components are 0 -- 1 and 2 -- 3 (each over 2 qubits), while the cmap component is star-shaped (let's say it has 10 qubits, with qubit 0 connected to every other qubit, but no other pairs of qubits are connected). Despite the fact that 2+2 < 10 (i.e. judging from the numbers of qubits alone it seems that it is possible to put both dag components on the cmap component), it is not actually possible to put both dag components on the cmap component (as qubit 0 would need to be used both times).

Reply:
We're not trying to find a perfect layout in this case, so if qubit 0 is the only one with connectivity it will be used with swaps across both components. This isn't any different than what happens today if you built a star coupling graph and passed it any circuit with connectivity between more than 2 qubits (ignoring the connected components).

The bigger concern for me, is in the case of multiple components is we want to minimize swaps induced by component selection/mapping (especially since afterwards `VF2PostLayout` will be able to select better performing qubits if there are any, and less swaps gives it more flexibility). The algorithm right now is fairly naive and I put it as a starting point as some that would work and we could improve further down the road (which is why I put a `TODO` comment inline about imprint the heuristic).

Reply:
> if qubit 0 is the only one with connectivity it will be used with swaps across both components

oh, I see this now, thanks!